examples/pytorch/graphsage/main.py

```diff
@@ -107,26 +78,127 @@ def create_dataset(
+        model.eval()
+
+    epochs_trained, steps_in_epoch_trained = load_checkpoint(
+        model, logger, args, session.get_world_rank()
+    )
+
+    optimizer = torch.optim.SGD(
         filter(lambda p: p.requires_grad, model.parameters()),
-        lr=args.learning_rate * world_size,
+        lr=args.learning_rate * session.get_world_size(),
+    )
+    optimizer = train.torch.prepare_optimizer(optimizer)
+
+    backend = create_backend(
+        BackendOptions(args), is_leader=(session.get_world_rank() == 0)
+    )
+    dataset = config["init_dataset_fn"](
+        args,
+        model,
+        rank=session.get_world_rank(),
+        world_size=session.get_world_size(),
+        backend=backend,
+    )
+    num_workers = (
+        0
+        if hasattr(dataset, "sampler_class")
+        and issubclass(dataset.sampler_class, (GENodeSampler, GEEdgeSampler))
+        or platform.system() == "Windows"
+        else args.data_parallel_num
+    )
+    dataset = torch.utils.data.DataLoader(
+        dataset=dataset,
+        num_workers=num_workers,
+    )
+    losses_full = []
+    for epoch in range(epochs_trained, args.num_epochs):
+        scores = []
+        labels = []
+        losses = []
+        for step, batch in enumerate(dataset):
+            if step < steps_in_epoch_trained:
+                continue
+            loss, score, label = model(batch)
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+
+            scores.append(score)
+            labels.append(label)
+            losses.append(loss.item())
+
+        losses_full.extend(losses)
+        steps_in_epoch_trained = 0
+        if epoch % args.save_ckpt_by_epochs == 0:
+            save_checkpoint(model, logger, epoch, step, args)
+
+        session.report(
+            {
+                "metric": model.compute_metric(scores, labels).item(),
+                "loss": np.mean(losses),
+                "losses": losses_full,
+            },
+        )
+
+
+def run_ray(init_dataset_fn, **kwargs):
+    """Run ray trainer."""
+    ray.init(num_cpus=3)
+
+    args = get_args(init_args, kwargs["run_args"] if "run_args" in kwargs else None)
+
+    trainer = TorchTrainer(
+        train_func,
+        train_loop_config={
+            "args": args,
+            "init_dataset_fn": init_dataset_fn,
+            **kwargs,
+        },
+        scaling_config=ScalingConfig(
+            num_workers=1, use_gpu=args.gpu, resources_per_worker={"CPU": 2}
+        ),
     )
+    return trainer.fit()
 
 
 def _main():
     # setup default logging component.
     setup_default_logging_config(enable_telemetry=True)
 
-    random.seed(42)
     # run_dist is the unified entry for pytorch model distributed training/evaluation/inference.
     # User only needs to prepare initializing function for model, dataset, optimizer and args.
     # reference: `deepgnn/pytorch/training/factory.py`
-    run_dist(
-        init_model_fn=create_model,
+    run_ray(
```

Comment:
why is [ray's example of pytorch-geometric](https://github.com/ray-project/ray/blob/8f8b95ed87090b9c2eb7b4916e760f1a29784497/python/ray/train/examples/pytorch_geometric/distributed_sage_example.py) is much shorter than this one(here code is split across multiple files of similar size)?

Reply:
Ours is 326 lines of code vs 230 lines. Most of this comes from the model being 126 vs their 30 lines, in the future we should use theirs, for now I keep the one we already have.